### PR TITLE
[ADD]open_academy: Added partner module with his view T#60931

### PR DESCRIPTION
**Pull request by UlisesRivadeneyra**
-Added open_academy_partner_views.xml
-Added sql constraints in open_cademy_course
-Rewrite of the copy function in open_academy_course
-Added new fields in open_academy_session
-Added new api decorators in open_academy_session(constraints and compute)
-Added new calendar view in open_academy_session_views
-Added color variation on fields of tree in open_academy<-session_views
-Added partner category in open_academy_partner_views

Im currently in the [Search views excersice ](https://www.odoo.com/documentation/15.0/developer/howtos/backend.html#id4)

### CODE REVIEW

**Comment by UlisesRivadeneyra** [comment]
@edy1192 Hi!!
This is my new PR i added the description above, its ready to be reviewed
Greeting, Ulises!!

File: open_academy/models/open_academy_course.py [hunk 1]
```diff
@@ -7,5 +7,20 @@ class Course(models.Model):
...
+    responsible_user_id = fields.Many2one("res.users", ondelete="set null")
+    session_ids = fields.One2many(
+        "openacademy.session", inverse_name="course_id")
 
-    responsible_user = fields.Many2one("res.users", ondelete="set null")
+    _sql_constraints = [
+        (
+            'name_description_check',
+            'CHECK(title != description)',
+            "The title of the course should not be the description"
```
Comment (edy1192) [inline open_academy/models/open_academy_course.py:18]:
```suggestion
            "The title of the course should not be the same that description."
```

File: open_academy/models/open_academy_course.py [hunk 2]
```diff
@@ -7,5 +7,20 @@ class Course(models.Model):
...
+        ),
+        ('name_unique', 'UNIQUE(title)', "The course title must be unique"),
```
Comment (edy1192) [inline open_academy/models/open_academy_course.py:20]:
```suggestion
        ('name_unique', 'UNIQUE(title)', "The course title must be unique."),
```

File: open_academy/models/open_academy_partner.py [hunk 3]
```diff
@@ -0,0 +1,8 @@
+from odoo import models, fields
+
+
+class Partner(models.Model):
```
Comment (edy1192) [inline open_academy/models/open_academy_partner.py:4]:
```suggestion
class ResPartner(models.Model):
```

File: open_academy/models/open_academy_session.py [hunk 4]
```diff
@@ -7,7 +8,75 @@ class Session(models.Model):
...
+    start_date = fields.Date(default=fields.Date.today)
+    active = fields.Boolean(default=True)
+    end_date = fields.Date(
+        store=True,
+        compute='_compute_get_end_date',
+        inverse='_inverse_set_end_date'
+    )
+
+    @api.depends('start_date', 'duration')
+    def _get_end_date(self):
```
Comment (edy1192) [inline open_academy/models/open_academy_session.py:34]:
```suggestion
    def _compute_end_date(self):
```

File: open_academy/models/open_academy_session.py [hunk 5]
```diff
@@ -7,7 +8,75 @@ class Session(models.Model):
...
+        ]
+    )
+    course_id = fields.Many2one('openacademy.course')
+    attendee_ids = fields.Many2many('res.partner')
+    taken_seats = fields.Float(compute='_compute_taken_seats')
+    start_date = fields.Date(default=fields.Date.today)
+    active = fields.Boolean(default=True)
+    end_date = fields.Date(
+        store=True,
+        compute='_compute_get_end_date',
```
Comment (edy1192) [inline open_academy/models/open_academy_session.py:29]:
```suggestion
        compute='_compute_end_date',
```

File: open_academy/models/open_academy_session.py [hunk 6]
```diff
@@ -7,7 +8,75 @@ class Session(models.Model):
...
+        inverse='_inverse_set_end_date'
```
Comment (edy1192) [inline open_academy/models/open_academy_session.py:30]:
```suggestion
        inverse='_inverse_end_date'
```

File: open_academy/models/open_academy_session.py [hunk 7]
```diff
@@ -7,7 +8,75 @@ class Session(models.Model):
...
+        for record in self:
+            if not (record.start_date and record.duration):
+                record.end_date = record.start_date
+                continue
+            start = fields.Datetime.from_string(record.start_date)
+            duration = timedelta(days=record.duration, seconds=-1)
+            record.end_date = start + duration
+
+    @api.depends('seats', 'attendee_ids')
+    def _taken_seats(self):
```
Comment (edy1192) [inline open_academy/models/open_academy_session.py:44]:
```suggestion
    def _taken_seats(self):_compute_taken_seats
```

File: open_academy/models/open_academy_session.py [hunk 8]
```diff
@@ -7,7 +8,75 @@ class Session(models.Model):
...
+        for record in self:
+            if not record.seats:
+                record.taken_seats = 0.0
+            else:
+                record.taken_seats = 100.0 * len(record.attendee_ids) / record.seats
```
Comment (edy1192) [inline open_academy/models/open_academy_session.py:49]:
```suggestion
                record.taken_seats = 100.0 * (len(record.attendee_ids) / record.seats)
```

File: open_academy/models/open_academy_session.py [hunk 9]
```diff
@@ -7,7 +8,75 @@ class Session(models.Model):
...
+
+    @api.onchange('seats', 'attendee_ids')
+    def _verify_valid_seats(self):
```
Comment (edy1192) [inline open_academy/models/open_academy_session.py:52]: Change this name, see the guidelines for onchange methods https://github.com/OCA/odoo-community.org/blob/master/website/Contribution/CONTRIBUTING.rst#37models

File: open_academy/models/open_academy_session.py [hunk 10]
```diff
@@ -7,7 +8,75 @@ class Session(models.Model):
...
+        if self.seats < 0:
+            return {
+                'warning': {
+                    'title': "Incorrect 'seats' value",
```
Comment (edy1192) [inline open_academy/models/open_academy_session.py:56]:
```suggestion
                    'title': "Incorrect 'seats' value.",
```

File: open_academy/models/open_academy_session.py [hunk 11]
```diff
@@ -7,7 +8,75 @@ class Session(models.Model):
...
+                    'message': "The number of available seats may not be negative",
```
Comment (edy1192) [inline open_academy/models/open_academy_session.py:57]:
```suggestion
                    'message': "The number of available seats may not be negative.",
```

File: open_academy/models/open_academy_session.py [hunk 12]
```diff
@@ -7,7 +8,75 @@ class Session(models.Model):
...
+                },
+            }
+        if self.seats < len(self.attendee_ids):
+            return {
+                'warning': {
+                    'title': "Too many attendees",
```
Comment (edy1192) [inline open_academy/models/open_academy_session.py:63]:
```suggestion
                    'title': "Too many attendees.",
```

File: open_academy/models/open_academy_session.py [hunk 13]
```diff
@@ -7,7 +8,75 @@ class Session(models.Model):
...
+                    'message': "Increase seats or remove excess attendees",
```
Comment (edy1192) [inline open_academy/models/open_academy_session.py:64]:
```suggestion
                    'message': "Increase seats or remove excess attendees.",
```

File: open_academy/models/open_academy_session.py [hunk 14]
```diff
@@ -7,7 +8,75 @@ class Session(models.Model):
...
+                },
+            }
+
+    @api.constrains('instructor_id', 'attendee_ids')
+    def _check_instructor_not_in_attendees(self):
+        for record in self:
+            if record.instructor_id in record.attendee_ids:
+                raise exceptions.ValidationError_("A session's instructor can't be an attendee")
```
Comment (edy1192) [inline open_academy/models/open_academy_session.py:72]: Why `_()`? That method is for Internationalization chapter.

File: open_academy/models/open_academy_session.py [hunk 15]
```diff
@@ -7,7 +8,75 @@ class Session(models.Model):
...
+            return {
+                'warning': {
+                    'title': "Too many attendees",
+                    'message': "Increase seats or remove excess attendees",
+                },
+            }
+
+    @api.constrains('instructor_id', 'attendee_ids')
+    def _check_instructor_not_in_attendees(self):
+        for record in self:
```
Comment (edy1192) [inline open_academy/models/open_academy_session.py:70]:
```suggestion
        for record in self.filtered('instructor_id'):
```

File: open_academy/views/open_academy_course_views.xml [hunk 16]
```diff
@@ -11,16 +11,14 @@
...
                         <field name="title"></field>
+                        <field name="description"></field>
+                        <field name="responsible_user_id"></field>
+                        <field name="session_ids"></field>
                     </group>
-                    <notebook>
-                        <page string="Description of the course">
-                            <field name="description"></field>
-                        </page>
-                    </notebook>
```
Comment (edy1192) [inline open_academy/views/open_academy_course_views.xml:23]: Why do you remove the description field?

File: open_academy/views/open_academy_menu_views.xml [hunk 17]
```diff
@@ -5,5 +5,6 @@
     parent='open_academy_root_menu' action="openacademy_course_action_list"/>
     <menuitem id="open_academy_session_menu" name="Sessions" sequence="2"
     parent='open_academy_root_menu' action="session_list_action"/>
-
+    <menuitem id="open_academy_partner_menu" name="Partner" sequence="3" 
+    parent="open_academy_root_menu" action="openacademy_partner_action_list"/>
```
Comment (edy1192) [inline open_academy/views/open_academy_menu_views.xml:9]: Remove this menu. This is not necessary, Odoo has an app for contacts. Find it in apps and install it.

File: open_academy/views/open_academy_partner_views.xml [hunk 18]
```diff
@@ -0,0 +1,34 @@
...
+        </field>
+    </record>
+
+    <record model="res.partner.category" id="teacher1">
+        <field name="name">Teacher / Level 1</field>
+    </record>
+
+    <record model="res.partner.category" id="teacher2">
+        <field name="name">Teacher / Level 2</field>
+    </record>
```
Comment (edy1192) [inline open_academy/views/open_academy_partner_views.xml:32]: Remove all this. This is not necessary, Odoo has an app for contacts. Find it in apps and install it.

File: open_academy/views/open_academy_partner_views.xml [hunk 19]
```diff
@@ -0,0 +1,34 @@
+<?xml version="1.0" encoding="utf-8"?>
+<odoo>
+
+    <record id="openacademy_partner_action_list" model="ir.actions.act_window">
+        <field name="name">Partners</field>
+        <field name="res_model">res.partner</field>
+        <field name="view_mode">tree,form</field>
+    </record>
```
Comment (edy1192) [inline open_academy/views/open_academy_partner_views.xml:8]: Remove all this. This is not necessary, Odoo has an app for contacts. Find it in apps and install it.

File: open_academy/views/open_academy_partner_views.xml [hunk 20]
```diff
@@ -0,0 +1,34 @@
...
+            <notebook>
+                <page string="Sessions">
+                    <group>
+                        <field name="instructor"></field>
+                        <field name="sessions_ids"/>
+                    </group>
+                </page>
+            </notebook>
+        </field>
+    </record>
```
Comment (edy1192) [inline open_academy/views/open_academy_partner_views.xml:24]: This is no correct. You need to inherit the `base.view_partner_form` and use xpath.

File: open_academy/views/open_academy_course_views.xml [hunk 21]
```diff
@@ -49,4 +48,4 @@
         </field>
     </record>
```
Comment (edy1192) [inline open_academy/views/open_academy_course_views.xml:50]: Where is the search views exercise? 

```
 Exercise

Search views

Add a button to filter the courses for which the current user is the responsible in the course search view. Make it selected by default.

Add a button to group courses by responsible user.
```

**Review by edy1192 — CHANGES_REQUESTED** [review]

**Comment by UlisesRivadeneyra** [comment]
Hi @edy1192 i already added the change in the code, hope that you can give it a check
Greetings, Ulises!!

File: open_academy/models/res_partner.py [hunk 22]
```diff
@@ -0,0 +1,15 @@
...
+    _inherit = "res.partner"
+
+    instructor = fields.Boolean()
+    sessions_ids = fields.Many2many("openacademy.session")
+    instructor_category = fields.Selection(
+        [
+            ('t_lvl1', 'Teacher / Level 1'),
+            ('t_lvl2', 'Teacher / Level 2'),
+        ]
+    )
```
Comment (edy1192) [inline open_academy/models/res_partner.py:14]: What is this? 

This exercise is completed with the functional explanation that I gave you about the categories.

Exercise says:

```
More complex domains

Create new partner categories Teacher / Level 1 and Teacher / Level 2. The instructor for a session can be either an instructor or a teacher (of any level).
```

File: open_academy/views/open_academy_course_views.xml [hunk 23]
```diff
@@ -11,16 +11,14 @@
...
         <field name="model">openacademy.course</field>
         <field name="arch" type="xml">
-            <form string="Course Form">
+            <form>
                 <sheet>
                     <group>
                         <field name="title"></field>
+                        <field name="description"></field>
+                        <field name="responsible_user_id"></field>
+                        <field name="session_ids"></field>
```
Comment (edy1192) [inline open_academy/views/open_academy_course_views.xml:20]: Move this field to a notebook section please.

File: open_academy/views/open_academy_session_views.xml [hunk 24]
```diff
@@ -21,10 +21,37 @@
...
+                        <field name="active"/>
+                        <field name="start_date"/>
                     </group>
                 </sheet>
             </form>
         </field>
     </record> 
 
+    <record model="ir.ui.view" id="session_view_tree">
+        <field name="name">openacademy.session_view_tree</field>
```
Comment (edy1192) [inline open_academy/views/open_academy_session_views.xml:35]:
```suggestion
        <field name="name">openacademy.session.tree.view</field>
```

File: open_academy/views/open_academy_course_views.xml [hunk 25]
```diff
@@ -32,7 +30,8 @@
         <field name="model">openacademy.course</field>
         <field name="arch" type="xml">
             <tree>
-                <field name="title"></field>
+                <field name="title"/>
+                <field name ="description"/>
```
Comment (edy1192) [inline open_academy/views/open_academy_course_views.xml:34]:
```suggestion
                <field name="description"/>
```

File: open_academy/views/res_partner_views.xml [hunk 26]
```diff
@@ -0,0 +1,30 @@
...
+            <xpath expr="//notebook" position="inside">
+                <page name="openacademy" string="Open Academy">
+                    <group name="instructor_info" string="Instructor Info" col='2'>
+                        <div colspan='1'>
+                            <label for="instructor" />
+                            <field name="instructor" />
+                        </div>
+                        <div colspan='1'>
+                            <label for='instructor_category' />
+                            <field name="instructor_category" />
```
Comment (edy1192) [inline open_academy/views/res_partner_views.xml:18]: This field is not necessary.

File: open_academy/models/open_academy_session.py (hunk 5 above)
Comment (edy1192) [reply open_academy/models/open_academy_session.py:29]:
```suggestion
        compute='_compute_end_date',
```

File: open_academy/models/open_academy_session.py (hunk 6 above)
Comment (edy1192) [reply open_academy/models/open_academy_session.py:30]:
```suggestion
        inverse='_inverse_end_date'
```

File: open_academy/models/open_academy_session.py (hunk 4 above)
Comment (edy1192) [reply open_academy/models/open_academy_session.py:34]: Pending https://github.com/UlisesRivadeneyra/Open-Academy/pull/4/files#r990351418

File: open_academy/models/open_academy_session.py (hunk 7 above)
Comment (edy1192) [reply open_academy/models/open_academy_session.py:44]: Pending https://github.com/UlisesRivadeneyra/Open-Academy/pull/4/files#r990352747

File: open_academy/models/open_academy_session.py (hunk 9 above)
Comment (edy1192) [reply open_academy/models/open_academy_session.py:52]: Pending https://github.com/UlisesRivadeneyra/Open-Academy/pull/4/files#r990354189

File: open_academy/models/open_academy_session.py (hunk 10 above)
Comment (edy1192) [reply open_academy/models/open_academy_session.py:56]: Pending https://github.com/UlisesRivadeneyra/Open-Academy/pull/4/files#r990356655

File: open_academy/models/open_academy_session.py (hunk 11 above)
Comment (edy1192) [reply open_academy/models/open_academy_session.py:57]: Same

File: open_academy/models/open_academy_session.py (hunk 12 above)
Comment (edy1192) [reply open_academy/models/open_academy_session.py:63]: Same

File: open_academy/models/open_academy_session.py (hunk 13 above)
Comment (edy1192) [reply open_academy/models/open_academy_session.py:64]: Same

File: open_academy/models/open_academy_session.py (hunk 15 above)
Comment (edy1192) [reply open_academy/models/open_academy_session.py:70]: Pending

File: open_academy/models/open_academy_session.py (hunk 14 above)
Comment (edy1192) [reply open_academy/models/open_academy_session.py:72]: ???

File: open_academy/models/open_academy_session.py [hunk 27]
```diff
@@ -7,7 +8,74 @@ class Session(models.Model):
...
+    attendee_ids = fields.Many2many('res.partner')
+    taken_seats = fields.Float(compute='_compute_taken_seats')
+    start_date = fields.Date(default=fields.Date.today)
+    active = fields.Boolean(default=True)
+    end_date = fields.Date(
+        store=True,
+        compute='_compute_get_end_date',
+        inverse='_inverse_set_end_date'
+    )
+
```
Comment (edy1192) [inline open_academy/models/open_academy_session.py:32]: Please organize the methods. See the guidelines https://github.com/OCA/odoo-community.org/blob/master/website/Contribution/CONTRIBUTING.rst#37models in `In a Model attribute order should be`

**Review by edy1192 — CHANGES_REQUESTED** [review]

File: open_academy/views/open_academy_session_views.xml (hunk 24 above)
Comment (UlisesRivadeneyra) [reply open_academy/views/open_academy_session_views.xml:35]: I have all names with _, are all names wrong?

Comment (edy1192) [reply open_academy/views/open_academy_session_views.xml:35]: Yes, please don't combine . and _. View's name is always with .

File: open_academy/views/open_academy_session_views.xml [hunk 28]
```diff
@@ -21,10 +21,37 @@
...
+                        <field name="taken_seats" widget="progressbar"/>
+                        <field name="active"/>
+                        <field name="start_date"/>
                     </group>
                 </sheet>
             </form>
         </field>
     </record> 
 
+    <record model="ir.ui.view" id="session_view_tree">
```
Comment (edy1192) [inline open_academy/views/open_academy_session_views.xml:34]: First ID then model. See the guidelines https://github.com/OCA/odoo-community.org/blob/master/website/Contribution/CONTRIBUTING.rst#252security-view-and-action
```suggestion
        <record id="session_view_tree" model="ir.ui.view" >
```

Please apply this suggestion in all records.

**Comment by UlisesRivadeneyra** [comment]
HI @edy1192 i added the modifications that you request.
Greetings, ulises

File: open_academy/models/open_academy_session.py [hunk 29]
```diff
@@ -1,13 +1,81 @@
...
+    @api.depends('seats', 'attendee_ids')
+    def _compute_taken_seats(self):
+        for record in self:
+            if not record.seats:
+                record.taken_seats = 0.0
+            else:
+                record.taken_seats = 100.0 * len(record.attendee_ids) / record.seats
+
+    @api.onchange('seats', 'attendee_ids')
+    def _onchange_valid_seats(self):
```
Comment (edy1192) [inline open_academy/models/open_academy_session.py:60]: Guidelines says: Onchange method: the onchange method pattern is `_onchange_<field_name>`

This should be:
```suggestion
    def _onchange_seats(self):
```

**Review by edy1192 — CHANGES_REQUESTED** [review]

**Comment by UlisesRivadeneyra** [comment]
Hi @edy1192 y change the last comment that you give me

**Comment by UlisesRivadeneyra** [comment]
Hi @edy1192 i just resolve the last comment

File: open_academy/models/open_academy_session.py [hunk 30]
```diff
@@ -1,13 +1,94 @@
...
+        if self.seats < len(self.attendee_ids):
+            return {
+                'warning': {
+                    'title': "Too many attendees.",
+                    'message': "Increase seats or remove excess attendees.",
+                },
+            }
+
+    @api.constrains('instructor_id', 'attendee_ids')
+    def _check_instructor_not_in_attendees(self):
```
Comment (edy1192) [inline open_academy/models/open_academy_session.py:90]: Constraints methods go before onchanges. 

```
In a Model attribute order should be
Private attributes (_name, _description, _inherit, ...)
Fields declarations
SQL constraints
Default method and _default_get
Compute and search methods in the same order than field declaration
**Constrains methods (@api.constrains) and onchange methods (@api.onchange)**
CRUD methods (ORM overrides)
Action methods
And finally, other business methods.
```

File: open_academy/models/open_academy_session.py [hunk 31]
```diff
@@ -1,13 +1,94 @@
...
+    color = fields.Integer()
+    course_id = fields.Many2one('openacademy.course')    
+    duration = fields.Integer()
+    end_date = fields.Date(
+        store=True,
+        compute='_compute_end_date',
+        inverse='_inverse_end_date'
+    )
+    hours = fields.Float(string="Duration in hours",
+                         compute='_get_hours', inverse='_set_hours')
```
Comment (edy1192) [inline open_academy/models/open_academy_session.py:22]: Where is the compute method and inverse method? And I don't see where you use this field?

File: open_academy/models/open_academy_session.py (hunk 8 above)
Comment (edy1192) [reply open_academy/models/open_academy_session.py:70]: You had applied this change, why did you remove it?

**Review by edy1192 — CHANGES_REQUESTED** [review]

**Comment by UlisesRivadeneyra** [comment]
Hi @edy1192 i already ordered the session.py file like the guideines says and i add the hours in the gant view, i also add the parentesis in the taken seats.

**Review by edy1192 — APPROVED** [review]

**Pull request merged**